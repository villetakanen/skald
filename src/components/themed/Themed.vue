<template>
  <div :class="`skald-themed ${theme}`">
    <!-- default background -->
    <div
      id="cover"
      :style="`height: ${posterHeight}px`"
      v-if="!poster">
      <div id="posterFadeOut"
        :style="`min-height: ${posterHeight/4}px`"></div>
    </div>
    <!-- site background -->
    <div
      id="cover"
      v-if="poster"
      :style="`height: ${posterHeight}px; background-image:url('${posterImageURL}')`">
      <div id="posterFadeOut"
        :style="`min-height: ${posterHeight/4}px`"></div>
    </div>

    <slot></slot>
    <!--div id="debug">
      siteid: {{ siteid }} <br/>
      Theme: {{ theme }} <br/>
      Poster: {{ poster }} <br/>
      PosterImageURL: {{ posterImageURL }} <br/>
      PosterHeight: {{ posterHeight }}
    </div-->
  </div>
</template>

<script>
import Vue from 'vue'
import VueCompositionApi, { defineComponent, ref, watch, onMounted, computed } from '@vue/composition-api'
import firebase from 'firebase/app'
import 'firebase/firestore'
import { useParams } from '../../lib/useParams'
import { fireStoreURL } from '../../plugins/skaldfire'
Vue.use(VueCompositionApi)

export default defineComponent({
  setup (props) {
    const theme = ref('')
    const poster = ref('')
    const siteid = ref('')
    const posterImageURL = ref('')
    const posterHeight = ref(window.innerHeight)

    function setTheme (params) {
      siteid.value = params.siteid

      if (params.siteid) {
        const db = firebase.firestore()
        const siteRef = db.collection('sites').doc(params.siteid)
        siteRef.get().then((doc) => {
          if (doc.exists) {
            theme.value = doc.data().theme
            const p = doc.data().posterURL
            poster.value = p
            if (p) {
              fireStoreURL(params.siteid + '/' + p).then((url) => {
                if (typeof url === 'string') posterImageURL.value = url
              })
            }
          }
        })
      } else {
        theme.value = ''
        poster.value = ''
      }
    }
    // useParams, and set theme based on the site opened
    const params = useParams()
    // reload/remount etc
    onMounted(() => { setTheme(params.value) })
    // route changed
    watch(params, setTheme)

    return { siteid, theme, poster, posterImageURL, posterHeight }
  }
})
</script>

<style lang="scss" scoped>
#debug{
  position: absolute;
  top: 16px;
  right: 16px;
  height: 128px;
  width: 256px;
  opacity: 0.75;
  background-color: black;
  color: greenyellow;
  z-index: 900000;
  padding: 8px;
}
// Cover image positioning and scaffold
div#cover{
  position: absolute;
  width: 100%;
  min-height: 300px;
  background-image: url('../../assets/skald-poster.svg');
  background-size: cover;
}
div#posterFadeOut{
  position: absolute;
  width: 100%;
  bottom: 0;
  background: linear-gradient(0deg, rgba(35,35,35,1) 0%, rgba(35,35,35,0) 100%);
}
</style>
