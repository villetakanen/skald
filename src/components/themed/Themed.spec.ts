import { mount } from '@vue/test-utils'
import Themed from './Themed.vue'

describe('Mounted App', () => {
  const wrapper = mount(Themed)

  test('is a Vue instance', () => {
    expect(wrapper.find('div').vm).toBeTruthy()
  })
})
