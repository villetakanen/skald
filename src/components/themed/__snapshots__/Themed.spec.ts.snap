// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounted App renders correctly 1`] = `
<div
  class="skald-themed "
>
  <div
    id="cover"
    style="height: 768px;"
  >
    <div
      id="posterFadeOut"
      style="min-height: 192px;"
    />
  </div>
   
  <!---->
   
</div>
`;
